File: go/console/output/error_test.go
package output_test

import (
	"bytes"
	"encoding/json"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"gopkg.in/yaml.v3"
	"hop.top/kit/go/console/output"
)

func TestRenderError_Table(t *testing.T) {
	var buf bytes.Buffer
	err := &output.Error{
		Code:         output.CodeNotFound,
		Message:      "thing missing",
		SuggestedFix: "create it first",
		ExitCode:     3,
	}
	require.NoError(t, output.RenderError(&buf, output.Table, err))
	got := buf.String()
	assert.Contains(t, got, "NOT_FOUND: thing missing")
	assert.Contains(t, got, "Fix: create it first")
}

func TestRenderError_EmptyFormatIsPlain(t *testing.T) {
	var buf bytes.Buffer
	err := &output.Error{
		Code:     output.CodeGeneric,
		Message:  "boom",
		ExitCode: 1,
	}
	require.NoError(t, output.RenderError(&buf, "", err))
	got := buf.String()
	assert.Contains(t, got, "GENERIC: boom")
	// No JSON braces in plain mode.
	assert.False(t, strings.Contains(got, "{"))
}

func TestRenderError_JSON(t *testing.T) {
	var buf bytes.Buffer
	err := &output.Error{
		Code:         output.CodeConflict,
		Message:      "already exists",
		Cause:        "duplicate key",
		SuggestedFix: "use a unique name",
		Alternatives: []string{"foo-2", "foo-3"},
		ExitCode:     4,
	}
	require.NoError(t, output.RenderError(&buf, output.JSON, err))

	var got output.Error
	require.NoError(t, json.Unmarshal(buf.Bytes(), &got))
	assert.Equal(t, output.CodeConflict, got.Code)
	assert.Equal(t, "already exists", got.Message)
	assert.Equal(t, "duplicate key", got.Cause)
	assert.Equal(t, "use a unique name", got.SuggestedFix)
	assert.Equal(t, []string{"foo-2", "foo-3"}, got.Alternatives)
	assert.Equal(t, 4, got.ExitCode)
}

func TestRenderError_YAML(t *testing.T) {
	var buf bytes.Buffer
	err := &output.Error{
		Code:     output.CodeUnauthorized,
		Message:  "forbidden",
		ExitCode: 5,
	}
	require.NoError(t, output.RenderError(&buf, output.YAML, err))
	var got output.Error
	require.NoError(t, yaml.Unmarshal(buf.Bytes(), &got))
	assert.Equal(t, output.CodeUnauthorized, got.Code)
	assert.Equal(t, 5, got.ExitCode)
}

func TestRenderError_NilIsNoop(t *testing.T) {
	var buf bytes.Buffer
	require.NoError(t, output.RenderError(&buf, output.JSON, nil))
	assert.Empty(t, buf.String())
}

func TestSentinelConstructors(t *testing.T) {
	tests := []struct {
		name           string
		got            *output.Error
		wantCode       string
		wantExit       int
		wantTransience string
	}{
		{"Generic", output.GenericError("boom"), output.CodeGeneric, 1, output.TransiencePermanent},
		{"NotFound", output.NotFoundError("nope"), output.CodeNotFound, 3, output.TransiencePermanent},
		{"Conflict", output.ConflictError("dup"), output.CodeConflict, 4, output.TransiencePermanent},
		{"Unauthorized", output.UnauthorizedError("nope"), output.CodeUnauthorized, 5, output.TransiencePermanent},
		{"Usage", output.UsageError("bad flag"), output.CodeUsage, 2, output.TransiencePermanent},
		{"RateLimited", output.RateLimitedError("budget"), output.CodeRateLimited, 64, output.TransienceTransient},
		{"Transient", output.TransientError("upstream timeout"), output.CodeTransient, 6, output.TransienceTransient},
		{"ProvenanceMissing", output.ProvenanceMissingError("/email"), output.CodeProvenanceMissing, 65, output.TransiencePermanent},
		{"ConsentRefused", output.ConsentRefusedError("refused"), output.CodeConsentRefused, 7, output.TransienceTransient},
		{"Prerequisite", output.PrerequisiteError("postgres unreachable"), output.CodePrerequisite, 70, output.TransienceTransient},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			require.NotNil(t, tc.got)
			assert.Equal(t, tc.wantCode, tc.got.Code)
			assert.Equal(t, tc.wantExit, tc.got.ExitCode)
			assert.Equal(t, tc.wantTransience, tc.got.Transience)
		})
	}
}

func TestGenericError(t *testing.T) {
	// Exit 1 is the catch-all class; the constructor closes the table so
	// adopters stop hand-rolling {ExitCode: 1} literals with no
	// Transience.
	assert.Equal(t, 1, output.ExitGeneric)

	e := output.GenericError("cache corrupt")
	require.NotNil(t, e)
	assert.Equal(t, output.CodeGeneric, e.Code)
	assert.Equal(t, output.ExitGeneric, e.ExitCode)
	assert.Equal(t, output.TransiencePermanent, e.Transience)
	assert.Equal(t, "cache corrupt", e.Message)
	assert.Equal(t, "GENERIC: cache corrupt", e.Error())
	assert.Nil(t, e.Unwrap())
}

func TestConsentRefusedError(t *testing.T) {
	// Exit 7 sits just past the spec's core 0-6 band. Pinned as a
	// literal so a later renumber has to change this line deliberately.
	assert.Equal(t, 7, output.ExitConsentRefused)
	assert.Equal(t, "CONSENT_REFUSED", output.CodeConsentRefused)

	e := output.ConsentRefusedError("refused: --confirm=no")
	require.NotNil(t, e)
	assert.Equal(t, output.CodeConsentRefused, e.Code)
	assert.Equal(t, output.ExitConsentRefused, e.ExitCode)
	assert.Equal(t, "refused: --confirm=no", e.Message)
	assert.Equal(t, "CONSENT_REFUSED: refused: --confirm=no", e.Error())

	// The whole point of the code: transient by class default, with no
	// WithTransience override at the call site. A caller clears it by
	// re-invoking with --confirm=yes.
	assert.Equal(t, output.TransienceTransient, e.Transience)
	assert.Equal(t, output.TransienceTransient,
		output.TransienceForCode(output.CodeConsentRefused))
}

func TestConsentRefusedIsDistinctFromUnauthorized(t *testing.T) {
	// Guards the separation this code exists to make. An auth or policy
	// denial is permanent; a declined confirmation is not. If these ever
	// collapse back onto one code, exit number or transience class, an
	// agent can no longer tell "never retry" from "retry with
	// --confirm=yes" on $? alone.
	consent := output.ConsentRefusedError("declined")
	auth := output.UnauthorizedError("no credentials")

	assert.NotEqual(t, auth.Code, consent.Code)
	assert.NotEqual(t, auth.ExitCode, consent.ExitCode)
	assert.NotEqual(t, auth.Transience, consent.Transience)

	assert.Equal(t, output.TransiencePermanent, auth.Transience)
	assert.Equal(t, output.TransienceTransient, consent.Transience)
}

func TestPrerequisiteError(t *testing.T) {
	// Exit 70 continues kit's contiguous extension band (64-69 are
	// taken). Pinned as a literal so a later renumber has to change
	// this line deliberately.
	assert.Equal(t, 70, output.ExitPrerequisite)
	assert.Equal(t, "PREREQUISITE", output.CodePrerequisite)

	e := output.PrerequisiteError("postgres unreachable at 127.0.0.1:5432")
	require.NotNil(t, e)
	assert.Equal(t, output.CodePrerequisite, e.Code)
	assert.Equal(t, output.ExitPrerequisite, e.ExitCode)
	assert.Equal(t, "postgres unreachable at 127.0.0.1:5432", e.Message)
	assert.Equal(t, "PREREQUISITE: postgres unreachable at 127.0.0.1:5432", e.Error())

	// Transient by class default, no WithTransience at the call site:
	// the operator starts the dependency and the identical command
	// succeeds.
	assert.Equal(t, output.TransienceTransient, e.Transience)
	assert.Equal(t, output.TransienceTransient,
		output.TransienceForCode(output.CodePrerequisite))
}

func TestPrerequisiteIsDistinctFromGenericAndTransient(t *testing.T) {
	// Guards the two separations this code exists to make.
	//
	// vs GENERIC: GENERIC means stop, the failure is uncharacterized.
	// PREREQUISITE means the invocation was correct and kit's logic
	// never ran — repair the environment and re-run verbatim. If these
	// collapse, an agent cannot tell "escalate" from "start the
	// dependency" on $? alone.
	//
	// vs TRANSIENT: TRANSIENT says a retry may clear this on its own.
	// A stopped dependency never comes up on its own, so a backoff loop
	// burns its budget and fails. Same transience class, deliberately
	// different code and exit number.
	prereq := output.PrerequisiteError("nothing listening")
	generic := output.GenericError("store corrupt")
	transient := output.TransientError("upstream timeout")

	assert.NotEqual(t, generic.Code, prereq.Code)
	assert.NotEqual(t, generic.ExitCode, prereq.ExitCode)
	assert.NotEqual(t, generic.Transience, prereq.Transience)

	assert.NotEqual(t, transient.Code, prereq.Code)
	assert.NotEqual(t, transient.ExitCode, prereq.ExitCode)
	// Transience intentionally agrees with TRANSIENT; the exit code is
	// what separates them.
	assert.Equal(t, transient.Transience, prereq.Transience)
}

func TestExtensionBandSlotsAreUnique(t *testing.T) {
	// kit allocates its extension band contiguously so no two features
	// claim the same slot. 64-69 are spoken for (RATE_LIMITED,
	// PROVENANCE_MISSING here; LEAK_DETECTED, CONFIG in
	// go/console/cli/conformance; GRADE_FAIL, GRADE_UNGRADABLE in
	// go/conformance/client). PREREQUISITE takes 70.
	//
	// The literals for the codes owned by other trees are repeated
	// rather than imported: importing them here would make the console
	// leaf depend on the conformance trees. This test's job is to fail
	// if a future allocation lands on a slot already in use.
	band := map[int]string{}
	for _, a := range []struct {
		exit int
		name string
	}{
		{output.ExitRateLimited, "RATE_LIMITED"},
		{output.ExitProvenanceMissing, "PROVENANCE_MISSING"},
		{66, "LEAK_DETECTED"},
		{67, "CONFIG"},
		{68, "GRADE_FAIL"},
		{69, "GRADE_UNGRADABLE"},
		{output.ExitPrerequisite, "PREREQUISITE"},
	} {
		if prior, dup := band[a.exit]; dup {
			t.Fatalf("exit %d claimed by both %s and %s", a.exit, prior, a.name)
		}
		band[a.exit] = a.name
	}
}

func TestTransienceForCode(t *testing.T) {
	tests := []struct {
		code string
		want string
	}{
		{output.CodeUsage, output.TransiencePermanent},
		{output.CodeNotFound, output.TransiencePermanent},
		{output.CodeConflict, output.TransiencePermanent},
		{output.CodeUnauthorized, output.TransiencePermanent},
		{output.CodeProvenanceMissing, output.TransiencePermanent},
		{output.CodeRateLimited, output.TransienceTransient},
		{output.CodeTransient, output.TransienceTransient},
		{output.CodeConsentRefused, output.TransienceTransient},
		{output.CodePrerequisite, output.TransienceTransient},
		{output.CodeGeneric, output.TransienceUnknown},
		{"ADOPTER_SPECIFIC", output.TransienceUnknown},
		{"", output.TransienceUnknown},
	}
	for _, tc := range tests {
		t.Run(tc.code, func(t *testing.T) {
			assert.Equal(t, tc.want, output.TransienceForCode(tc.code))
		})
	}
}

func TestWrapError_DefaultsTransienceFromCode(t *testing.T) {
	base := assert.AnError
	assert.Equal(t, output.TransiencePermanent,
		output.WrapError(base, output.CodeConflict, 4).Transience)
	assert.Equal(t, output.TransienceTransient,
		output.WrapError(base, output.CodeRateLimited, 64).Transience)
	assert.Equal(t, output.TransienceUnknown,
		output.WrapError(base, output.CodeGeneric, 1).Transience)
}

func TestWithTransience_CopiesAndSets(t *testing.T) {
	orig := &output.Error{Code: "SHARED", Message: "m", ExitCode: 9}
	got := orig.WithTransience(output.TransienceTransient)
	require.NotNil(t, got)
	assert.NotSame(t, orig, got)
	assert.Equal(t, output.TransienceTransient, got.Transience)
	// Shared package-level envelopes must never be mutated in place.
	assert.Empty(t, orig.Transience)
	// Every other rendered field carries over.
	assert.Equal(t, orig.Code, got.Code)
	assert.Equal(t, orig.Message, got.Message)
	assert.Equal(t, orig.ExitCode, got.ExitCode)

	var nilErr *output.Error
	assert.Nil(t, nilErr.WithTransience(output.TransienceTransient))
}

func TestRenderError_StructuredAlwaysCarriesTransience(t *testing.T) {
	// A literal built without Transience must still render a valid
	// transience class in structured formats (spec Factor 4: every
	// structured error carries transient|permanent|unknown).
	var buf bytes.Buffer
	require.NoError(t, output.RenderError(&buf, output.JSON, &output.Error{
		Code: "ADOPTER_SPECIFIC", Message: "m", ExitCode: 9,
	}))
	var got map[string]any
	require.NoError(t, json.Unmarshal(buf.Bytes(), &got))
	assert.Equal(t, output.TransienceUnknown, got["transience"])

	buf.Reset()
	require.NoError(t, output.RenderError(&buf, output.YAML, &output.Error{
		Code: "ADOPTER_SPECIFIC", Message: "m", ExitCode: 9,
	}))
	assert.Contains(t, buf.String(), "transience: unknown")

	// An explicit class renders untouched.
	buf.Reset()
	require.NoError(t, output.RenderError(&buf, output.JSON,
		output.RateLimitedError("budget")))
	got = nil
	require.NoError(t, json.Unmarshal(buf.Bytes(), &got))
	assert.Equal(t, output.TransienceTransient, got["transience"])
}

func TestError_ImplementsErrorInterface(t *testing.T) {
	// *output.Error should satisfy the error interface so adopters can
	// return it directly from RunE.
	var _ error = (*output.Error)(nil)

	e := output.NotFoundError("missing thing")
	assert.Contains(t, e.Error(), "NOT_FOUND")
	assert.Contains(t, e.Error(), "missing thing")
}

func TestError_AsCLIError_RoundTrips(t *testing.T) {
	e := output.ConflictError("dup")
	got := e.AsCLIError()
	assert.Same(t, e, got)
}

File: sdk/ts/test/output/error.test.ts
import { describe, it, expect } from 'vitest';
import * as yaml from 'js-yaml';
import {
  CODE_CONFLICT,
  CODE_GENERIC,
  CODE_NOT_FOUND,
  CODE_PROVENANCE_MISSING,
  CODE_RATE_LIMITED,
  CODE_TRANSIENT,
  CODE_UNAUTHORIZED,
  CODE_USAGE,
  EXIT_GENERIC,
  EXIT_PROVENANCE_MISSING,
  EXIT_RATE_LIMITED,
  EXIT_TRANSIENT,
  TRANSIENCE_PERMANENT,
  TRANSIENCE_TRANSIENT,
  TRANSIENCE_UNKNOWN,
  conflictError,
  genericError,
  notFoundError,
  provenanceMissingError,
  rateLimitedError,
  renderError,
  transienceForCode,
  transientError,
  unauthorizedError,
  unwrapError,
  usageError,
  withTransience,
  wrapError,
} from '../../src/output/error';
import type { CliError } from '../../src/output/error';

/** Collects renderError output into a string. */
function capture(): { write(chunk: string): void; text(): string } {
  let buf = '';
  return {
    write(chunk: string) {
      buf += chunk;
    },
    text: () => buf,
  };
}

describe('constructors — code/exit/transience', () => {
  const cases: Array<[string, CliError, string, number, string]> = [
    ['Generic', genericError('boom'), CODE_GENERIC, 1, TRANSIENCE_PERMANENT],
    ['NotFound', notFoundError('nope'), CODE_NOT_FOUND, 3, TRANSIENCE_PERMANENT],
    ['Conflict', conflictError('dup'), CODE_CONFLICT, 4, TRANSIENCE_PERMANENT],
    [
      'Unauthorized',
      unauthorizedError('nope'),
      CODE_UNAUTHORIZED,
      5,
      TRANSIENCE_PERMANENT,
    ],
    ['Usage', usageError('bad flag'), CODE_USAGE, 2, TRANSIENCE_PERMANENT],
    [
      'RateLimited',
      rateLimitedError('budget'),
      CODE_RATE_LIMITED,
      64,
      TRANSIENCE_TRANSIENT,
    ],
    [
      'Transient',
      transientError('upstream timeout'),
      CODE_TRANSIENT,
      6,
      TRANSIENCE_TRANSIENT,
    ],
    [
      'ProvenanceMissing',
      provenanceMissingError('/email'),
      CODE_PROVENANCE_MISSING,
      65,
      TRANSIENCE_PERMANENT,
    ],
  ];
  it.each(cases)('%s', (_name, got, wantCode, wantExit, wantTransience) => {
    expect(got.code).toBe(wantCode);
    expect(got.exit_code).toBe(wantExit);
    expect(got.transience).toBe(wantTransience);
  });

  it('exit-code table is unique, 1 generic / 6 transient / 65 provenance', () => {
    const exits = new Map(cases.map(([, e]) => [e.exit_code, e.code]));
    expect(exits.size).toBe(8);
    expect(EXIT_GENERIC).toBe(1);
    expect(EXIT_TRANSIENT).toBe(6);
    expect(EXIT_RATE_LIMITED).toBe(64);
    expect(EXIT_PROVENANCE_MISSING).toBe(65);
    expect(exits.get(1)).toBe(CODE_GENERIC);
    expect(exits.get(6)).toBe(CODE_TRANSIENT);
    expect(exits.get(65)).toBe(CODE_PROVENANCE_MISSING);
  });
});

describe('transienceForCode', () => {
  it.each([
    [CODE_USAGE, TRANSIENCE_PERMANENT],
    [CODE_NOT_FOUND, TRANSIENCE_PERMANENT],
    [CODE_CONFLICT, TRANSIENCE_PERMANENT],
    [CODE_UNAUTHORIZED, TRANSIENCE_PERMANENT],
    [CODE_PROVENANCE_MISSING, TRANSIENCE_PERMANENT],
    [CODE_RATE_LIMITED, TRANSIENCE_TRANSIENT],
    [CODE_TRANSIENT, TRANSIENCE_TRANSIENT],
    [CODE_GENERIC, TRANSIENCE_UNKNOWN],
    ['ADOPTER_SPECIFIC', TRANSIENCE_UNKNOWN],
    ['', TRANSIENCE_UNKNOWN],
  ])('%s -> %s', (code, want) => {
    expect(transienceForCode(code)).toBe(want);
  });
});

describe('wrapError', () => {
  it('defaults transience from code', () => {
    const base = new Error('boom');
    expect(wrapError(base, CODE_CONFLICT, 4)?.transience).toBe(
      TRANSIENCE_PERMANENT,
    );
    expect(wrapError(base, CODE_RATE_LIMITED, 64)?.transience).toBe(
      TRANSIENCE_TRANSIENT,
    );
    expect(wrapError(base, CODE_GENERIC, 1)?.transience).toBe(
      TRANSIENCE_UNKNOWN,
    );
  });

  it('retains the source error and passes null through', () => {
    const base = new Error('boom');
    const e = wrapError(base, CODE_CONFLICT, 4);
    expect(e?.message).toBe('boom');
    expect(unwrapError(e as CliError)).toBe(base);
    expect(wrapError(null, CODE_GENERIC, 1)).toBeNull();
  });
});

describe('withTransience', () => {
  it('copies and sets, never mutates the shared envelope', () => {
    const orig: CliError = { code: 'SHARED', message: 'm', exit_code: 9 };
    const got = withTransience(orig, TRANSIENCE_TRANSIENT);
    expect(got).not.toBe(orig);
    expect(got.transience).toBe(TRANSIENCE_TRANSIENT);
    expect(orig.transience).toBeUndefined();
    expect(got.code).toBe(orig.code);
    expect(got.message).toBe(orig.message);
    expect(got.exit_code).toBe(orig.exit_code);
  });

  it('carries the retained source error to the copy', () => {
    const base = new Error('boom');
    const e = wrapError(base, CODE_GENERIC, 1) as CliError;
    expect(unwrapError(withTransience(e, TRANSIENCE_PERMANENT))).toBe(base);
  });
});

describe('renderError — structured always carries transience', () => {
  it('normalizes unset transience to unknown in JSON', () => {
    const w = capture();
    renderError(w, 'json', { code: 'ADOPTER_SPECIFIC', message: 'm', exit_code: 9 });
    expect(JSON.parse(w.text()).transience).toBe(TRANSIENCE_UNKNOWN);
  });

  it('normalizes unset transience to unknown in YAML', () => {
    const w = capture();
    renderError(w, 'yaml', { code: 'ADOPTER_SPECIFIC', message: 'm', exit_code: 9 });
    expect(w.text()).toContain('transience: unknown');
  });

  it('renders an explicit class untouched', () => {
    const w = capture();
    renderError(w, 'json', rateLimitedError('budget'));
    expect(JSON.parse(w.text()).transience).toBe(TRANSIENCE_TRANSIENT);
  });

  it('does not mutate the input envelope', () => {
    const e: CliError = { code: 'ADOPTER_SPECIFIC', message: 'm', exit_code: 9 };
    renderError(capture(), 'json', e);
    expect(e.transience).toBeUndefined();
  });
});

describe('renderError — wire round-trip', () => {
  it('JSON: full envelope, empty optionals off the wire', () => {
    let w = capture();
    renderError(w, 'json', provenanceMissingError('/email'));
    const full = JSON.parse(w.text());
    expect(full.code).toBe(CODE_PROVENANCE_MISSING);
    expect(full.cause).toBe('/email');
    expect(full.exit_code).toBe(65);
    expect(full.transience).toBe(TRANSIENCE_PERMANENT);

    w = capture();
    renderError(w, 'json', transientError('upstream timeout'));
    const bare = JSON.parse(w.text());
    expect(Object.keys(bare).sort()).toEqual([
      'code',
      'exit_code',
      'message',
      'transience',
    ]);
    expect(bare.exit_code).toBe(6);
  });

  it('YAML: parses back to the wire shape', () => {
    const w = capture();
    renderError(w, 'yaml', transientError('upstream timeout'));
    expect(yaml.load(w.text())).toEqual({
      code: CODE_TRANSIENT,
      message: 'upstream timeout',
      exit_code: 6,
      transience: TRANSIENCE_TRANSIENT,
    });
  });
});

describe('renderError — plain text + null', () => {
  it('renders each populated field on its own line', () => {
    const w = capture();
    renderError(w, 'table', {
      code: 'NOT_FOUND',
      message: 'missing thing',
      cause: 'root',
      suggested_fix: 'try --all',
      alternatives: ['other'],
      exit_code: 3,
    });
    expect(w.text()).toBe(
      'NOT_FOUND: missing thing\nCause: root\nFix: try --all\nAlternative: other\n',
    );
  });

  it('renders bare message without code prefix', () => {
    const w = capture();
    renderError(w, '', { code: '', message: 'just text', exit_code: 1 });
    expect(w.text()).toBe('just text\n');
  });

  it('null renders nothing', () => {
    const w = capture();
    renderError(w, 'json', null);
    expect(w.text()).toBe('');
  });
});
